### ISSUE TITLE
Add rotary embeddings to MultiHeadAttention.

### ISSUE THREAD

**Issue by loua19** [pytorch/pytorch#97899]
### 🚀 The feature, motivation and pitch

Rotary embeddings are becoming more and more popular. It would be nice if PyTorch supported them directly, instead of having to refer to research libraries.

Rotary embeddings require performing transformations to key and query matrices as described in the following paper:

https://arxiv.org/abs/2104.09864

An example implementation can be found in the xFormers library. Due to the fact that these transformations are applied to the key and query matrices (i.e. after the projections) and not the hidden states, it would need to be added as an option for F.multi_head_attention.

cc @svekars @carljparker @jbschlosser @bhosmer @cpuhrsch @erichan1

**Comment by jbschlosser** [comment]
Possibly related: #24826

**Comment by loua19** [comment]
> Possibly related: #24826

This is related of course, however the main difference is that its possible to manually implement on-top of nn.MultIHeadAttention.

To illustrate the difference, most versions of positional embeddings (i.e. sinusoidal, learned, ect...) involve the step x := x +p, where x has shape (batch_size, sequence_length, d_model). This adds a position embedding directly to the hidden states. Therefore we can use it easily with nn.MultiHeadAttention by doing:

```
hidden_states_pos_enc = positional_encoding(hidden_states)  # i.e. x := x + p
hidden_states_next = multi_head_attention(hidden_states_pos_enc)
```

This contrasts with Rotary embeddings, where we apply a transformation to the key and query matrices. This happens inside nn.MultiHeadAttention as it requires the hidden states to have the key and query transformations applied to them first. Therefore there is no way (that I can see) to take advantage of nn.MultiHeadAttention (and the efficiency) while also using rotary embeddings. All implementations I've seen online (including my own) require you to reimplement multi-head attention yourself in order to use rotary embeddings.

**Comment by loua19** [comment]
I would be happy to contribute this myself, as I have a working version (see ManualEncoderBlock - https://github.com/loua19/muse/blob/main/muse/models/model.py). As this seems like a big feature however, it would probably be best for someone who knows the codebase better to implement it.

**Comment by jbschlosser** [comment]
Thanks for the additional clarification! This sounds like another case where our `nn.MultiheadAttention` module is too rigid. To solve issues like this, we've recently focused on providing performant building blocks rather than an opaque high-level module that has difficulty supporting the variations users need. In particular, check out our new work for fast scaled dot-product attention [here](https://pytorch.org/tutorials/intermediate/scaled_dot_product_attention_tutorial.html). I'm curious if using the SDPA we describe there instead of `nn.MultiheadAttention` directly would give you the flexibility you need?

**Comment by loua19** [comment]
This looks great! It seems like I can mess around with the key, query, and value matrices (after being projected from the hidden states) before providing them as arguments for attention. Thanks for getting back to me. 

Do I need to install the nightly to have access to this feature?

**Comment by loua19** [comment]
> Thanks for the additional clarification! This sounds like another case where our `nn.MultiheadAttention` module is too rigid. To solve issues like this, we've recently focused on providing performant building blocks rather than an opaque high-level module that has difficulty supporting the variations users need. In particular, check out our new work for fast scaled dot-product attention [here](https://pytorch.org/tutorials/intermediate/scaled_dot_product_attention_tutorial.html). I'm curious if using the SDPA we describe there instead of `nn.MultiheadAttention` directly would give you the flexibility you need?

I also have another question. Will the attention mask get broadcasted? It doesn't specify in the docs.

For instance, if I supply a padding mask (I am only using BERT style self attention) of shape (batch_size, 1, 1, seq_len), then will this get broadcast to (batch_size, num_heads, seq_len, seq_len) as it requires in the docs. When I run my code, it works, however I'm sceptical if the correct thing is happening behind the scenes.

Thanks again for your help.

**Comment by jbschlosser** [comment]
> Do I need to install the nightly to have access to this feature?

The feature is present in PyTorch >= 2.0.0, so nightly isn't a requirement.

> For instance, if I supply a padding mask (I am only using BERT style self attention) of shape (batch_size, 1, 1, seq_len), then will this get broadcast to (batch_size, num_heads, seq_len, seq_len) as it requires in the docs. When I run my code, it works, however I'm sceptical if the correct thing is happening behind the scenes.

Yes, it'll appropriately broadcast. We should probably clarify this in the docs.

A few other things to note here:
* `attn_mask` for SDPA is unfortunately defined opposite to that of `nn.MultiheadAttention`
    * Where `attn_mask=True`: SDPA **includes** for attention, while `nn.MultiheadAttention` **excludes**
    * `logical_not` can help here for converting an `attn_mask` usable for MHA to one usable for SDPA
* If you're using a standard causal `attn_mask` (i.e. item attention is restricted to previous items in the sequence only), you can get better perf by passing `is_causal=True` and skipping the `attn_mask` altogether. This is because we have optimizations for the causal case that avoid materializing the mask entirely.
* You can alternatively pass ["nested tensors"](https://pytorch.org/docs/stable/nested.html) for `q,k,v` through SDPA rather than using an `attn_mask`. Nested tensors can hold ragged inputs, preventing the need for padding + masking. This is another newer feature we're working on, and the two fit together well now. That said, op support is currently limited, so this might not be an option for you right now.

Feel free to ask any more questions you may have!

**Comment by jbschlosser** [comment]
Actionable to clarify `attn_mask` broadcasting in the docs.

**Comment by loua19** [comment]
> > Do I need to install the nightly to have access to this feature?
> 
> The feature is present in PyTorch >= 2.0.0, so nightly isn't a requirement.
> 
> > For instance, if I supply a padding mask (I am only using BERT style self attention) of shape (batch_size, 1, 1, seq_len), then will this get broadcast to (batch_size, num_heads, seq_len, seq_len) as it requires in the docs. When I run my code, it works, however I'm sceptical if the correct thing is happening behind the scenes.
> 
> Yes, it'll appropriately broadcast. We should probably clarify this in the docs.
> 
> A few other things to note here:
> 
> * `attn_mask` for SDPA is unfortunately defined opposite to that of `nn.MultiheadAttention`
>   
>   * Where `attn_mask=True`: SDPA **includes** for attention, while `nn.MultiheadAttention` **excludes**
>   * `logical_not` can help here for converting an `attn_mask` usable for MHA to one usable for SDPA
> * If you're using a standard causal `attn_mask` (i.e. item attention is restricted to previous items in the sequence only), you can get better perf by passing `is_causal=True` and skipping the `attn_mask` altogether. This is because we have optimizations for the causal case that avoid materializing the mask entirely.
> * You can alternatively pass ["nested tensors"](https://pytorch.org/docs/stable/nested.html) for `q,k,v` through SDPA rather than using an `attn_mask`. Nested tensors can hold ragged inputs, preventing the need for padding + masking. This is another newer feature we're working on, and the two fit together well now. That said, op support is currently limited, so this might not be an option for you right now.
> 
> Feel free to ask any more questions you may have!

**Comment by loua19** [comment]
Thanks for the information and clarification. I'm actually not using casual attention, I was trying to implement a equivilent key_padding_mask and I think I've done it correctly. 

By the way, using F.scaled_dot_product_attention() has given me 40% performance increase and 20% memory increase, so thanks for that.

On the topic of rotary embeddings, just in case it useful in the future. I have the following code which implements it with F.scaled_dot_product_attention(). Note the rotary code is taken from LLaMa.

```
# Taken from facebookresearch/llama/model.py
def reshape_for_broadcast(freqs_cis: torch.Tensor, x: torch.Tensor):
    ndim = x.ndim
    assert 0 <= 1 < ndim
    assert freqs_cis.shape == (x.shape[1], x.shape[-1])
    shape = [d if i == 1 or i == ndim - 1 else 1 for i, d in enumerate(x.shape)]
    return freqs_cis.view(*shape)


# Taken from facebookresearch/llama/model.py
def apply_rotary_emb(
    xq: torch.Tensor,
    xk: torch.Tensor,
    freqs_cis: torch.Tensor,
) -> Tuple[torch.Tensor, torch.Tensor]:
    xq_ = torch.view_as_complex(xq.float().reshape(*xq.shape[:-1], -1, 2))
    xk_ = torch.view_as_complex(xk.float().reshape(*xk.shape[:-1], -1, 2))
    freqs_cis = reshape_for_broadcast(freqs_cis, xq_)
    xq_out = torch.view_as_real(xq_ * freqs_cis).flatten(3)
    xk_out = torch.view_as_real(xk_ * freqs_cis).flatten(3)

    return xq_out.type_as(xq), xk_out.type_as(xk)

# Taken from facebookresearch/llama/model.py
def precompute_freqs_cis(dim: int, end: int, theta: float = 10000.0):
    freqs = 1.0 / (
        theta ** (torch.arange(0, dim, 2)[: (dim // 2)].float() / dim)
    )
    t = torch.arange(end, device=freqs.device)  # type: ignore
    freqs = torch.outer(t, freqs).float()  # type: ignore
    freqs_cis = torch.polar(torch.ones_like(freqs), freqs)  # complex64

    return freqs_cis


class FusedEncoderBlock(nn.Module):
    """Transformer encoder block using F.scaled_dot_product_attention().

    This block has the following changes from a typical transformer encoder:

        - Rotary embeddings are applied to the key/query matrices.
        - Layer norm is applied before attention and feed forward, instead of
            after.
        - Keys arising from padding are masked during attention.
        - GELU activation is used instead of ReLU.

    Args:
        model_config (ModelConfig): Model config settings.
    """

    def __init__(self, model_config: ModelConfig):
        super().__init__()

        self.pad_id = model_config.pad_id
        self.drop_p = model_config.drop_p
        self.n_heads = model_config.n_heads
        self.d_head = model_config.d_model // model_config.n_heads
        self.max_seq_len = model_config.max_seq_len

        # Attention
        self.q = nn.Linear(
            in_features=model_config.d_model,
            out_features=model_config.d_model,
            bias=False,
        )
        self.k = nn.Linear(
            in_features=model_config.d_model,
            out_features=model_config.d_model,
            bias=False,
        )
        self.v = nn.Linear(
            in_features=model_config.d_model,
            out_features=model_config.d_model,
            bias=False,
        )
        self.att_proj_linear = nn.Linear(
            in_features=model_config.d_model,
            out_features=model_config.d_model,
        )
        self.resid_dropout = nn.Dropout(model_config.drop_p)

        # FF Layer
        self.ff_dropout = nn.Dropout(model_config.drop_p)
        self.ff_linear_1 = nn.Linear(
            in_features=model_config.d_model,
            out_features=model_config.d_model * model_config.ff_mult,
        )
        self.ff_linear_2 = nn.Linear(
            in_features=model_config.d_model * model_config.ff_mult,
            out_features=model_config.d_model,
        )
        self.ff_activation = nn.GELU()

        # Pre layer norms
        self.norm1 = nn.LayerNorm(model_config.d_model)
        self.norm2 = nn.LayerNorm(model_config.d_model)

    def forward(
        self, x: torch.Tensor, pad_mask: torch.Tensor, freqs_cis: torch.Tensor
    ):
        x = x + self._att_block(self.norm1(x), pad_mask, freqs_cis)
        x = x + self._ff_block(self.norm2(x))

        return x

    def _att_block(
        self, x: torch.Tensor, pad_mask: torch.Tensor, freqs_cis: torch.Tensor
    ):
        batch_size, seq_len, _ = x.shape
        xq, xk, xv = self.q(x), self.k(x), self.v(x)

        # Reshape for rotary embeddings
        xq = xq.view(batch_size, seq_len, self.n_heads, self.d_head)
        xk = xk.view(batch_size, seq_len, self.n_heads, self.d_head)
        xv = xv.view(batch_size, seq_len, self.n_heads, self.d_head)
        xq, xk = apply_rotary_emb(xq, xk, freqs_cis)

        # Reshape for attention calculation: (b_sz, n_head, s_len, d_head)
        xq = xq.transpose(1, 2)
        xk = xk.transpose(1, 2)
        xv = xv.transpose(1, 2)

        # Required as we are not using a nn.Dropout layer
        if self.training:
            att_dropout = self.drop_p
        else:
            att_dropout = 0.0

        # Using beta torch functionality (subject to change)
        # See - https://shorturl.at/jtI17
        att = F.scaled_dot_product_attention(
            query=xq,
            key=xk,
            value=xv,
            attn_mask=pad_mask,
            dropout_p=att_dropout,
            is_causal=False,
        )

        # Shape (b_sz, s_len, n_head, d_head)
        out = att.transpose(1, 2).contiguous()
        out = out.view(batch_size, seq_len, self.n_heads * self.d_head)

        return self.resid_dropout(self.att_proj_linear(out))

    def _ff_block(self, x: torch.Tensor):
        x = self.ff_linear_2(self.ff_activation(self.ff_linear_1(x)))

        return self.ff_dropout(x)

```

**Comment by loua19** [comment]
> Feel free to ask any more questions you may have!

Just to be clear. If I am doing BERT style self attention, then should the padding attention mask I provide be of shape (batch, 1, 1, seq_len) or (batch, 1, seq_len, 1).

**Comment by cpuhrsch** [comment]
@loua19 - Have you tried applying `torch.compile` to `FusedEncoderBlock`?

**Comment by s-rog** [comment]
```
# pad_mask.shape: (batch_size, seq_len)
pad_mask = pad_mask.view(batch_size, 1, 1, seq_len)
pad_mask = pad_mask.expand(-1, n_heads, -1, -1)
pad_mask = pad_mask.view(batch_size, n_heads, -1, seq_len)
```

@loubbrad 
Not sure if this is what you're after but I added this to convert pad mask to attn mask for SDPA. Thanks for the code snippet.
